modin/pandas/test/test_series.py

```diff
@@ -4510,18 +4516,75 @@ def test_struct_general():
     )
 
 
-def test_case_when():
-    # Copied from pandas
-    c_md, c_pd = create_test_series([6, 7, 8, 9], name="c")
-    a_md, a_pd = create_test_series([0, 0, 1, 2])
-    b_md, b_pd = create_test_series([0, 3, 4, 5])
+def _case_when_caselists():
+    def permutations(values):
+        return [
+            p
+            for r in range(1, len(values) + 1)
+            for p in itertools.permutations(values, r)
+        ]
 
-    results = [None, None]
-    for idx, (c, a, b) in enumerate(((c_md, a_md, b_md), (c_pd, a_pd, b_pd))):
-        results[idx] = c.case_when(
-            caselist=[(a.gt(0), a), (b.gt(0), b)]  # condition, replacement
+    conditions = permutations(
+        [
+            [True, False, False, False] * 10,
+            pandas.Series([True, False, False, False] * 10),
+            pandas.Series([True, False, False, False] * 10, index=range(78, -2, -2)),
+            lambda df: df.gt(0),
+        ]
+    )
+    replacements = permutations([[0, 3, 4, 5] * 10, 0, lambda df: 1])
+    caselists = []
+    for c in conditions:
+        for r in replacements:
+            if len(c) == len(r):
+                caselists.append(list(zip(c, r)))
+    return caselists
+
+
+@pytest.mark.parametrize(
+    "base",
+    [
+        pandas.Series(range(40)),
+        pandas.Series([0, 7, 8, 9] * 10, name="c", index=range(0, 80, 2)),
+    ],
+)
+@pytest.mark.parametrize(
+    "caselist",
+    _case_when_caselists(),
+)
+def test_case_when(base, caselist):
+    pandas_result = base.case_when(caselist)
+    modin_bases = [pd.Series(base)]
+
+    # 'base' and serieses from 'caselist' must have equal lengths, however in this test we want
+    # to verify that 'case_when' works correctly even if partitioning of 'base' and 'caselist' isn't equal
+    if StorageFormat.get() != "Hdk":  # HDK always uses a single partition.
```

Comment:
we should better fix [this problem](https://github.com/modin-project/modin/pull/6972/files#r1521307926) at the test level
```suggestion
    if StorageFormat.get() != "Hdk" and StorageFormat.get() != "BaseOnPython":  # HDK and BaseOnPython always use a single partition.
```